10-Profiling.md

```diff
@@ -137,129 +141,182 @@ If you just want to see how long it takes your program to run, this is fine.
+}
+
+int main()
+{
+  const int count = 200;
+
+  do_iterative(count * 1000);
+
+  do_recursive(count);
+  return 0;
 }
 ~~~
 
-In order to use `gprof` we have to pass an additional flag to `g++`.
-The `-pg` flag tells `g++` to record profile information whenever our compiled program runs.
-Each time you run your program it will generate a file called `gmon.out` which contains
-information that can then be interpreted by the `gprof` command.
+In order to use `gperftool`, we have to link its library with our code using the `-lprofiler` flag.
+We'll also want to use the `-g` flag so that `kcachegrind` can show us our source code when we're viewing our profiling output!
 
 Let's compile the program above.
-We'll assume it's called `main.cpp`.
+We'll assume it's called `fibonnaci.cpp`.
 
 ~~~shell
 $ ls
-main.cpp
-$ g++ -pg -o main main.cpp
+fibonnaci.cpp
+$ g++ -g -lprofiler -o  fibonnaci.cpp
 $ ls
-main    main.cpp
+fibonnaci    fibonnaci.cpp
 ~~~
 
-In order to generate `gmon.out`, we need to run `main`.
+By default, `gperftool` doesn't do anything when you execute your program.
+To get it to profile your code, you will need to set an environment variable, `CPUPROFILE`, to the name of the file where you want to store your profile output.
+We'll store ours in a file named `gperftool.prof`:
 
 ~~~shell
+$ CPUPROFILE=gperftool.prof ./fibonnaci
+Iterative: 10946
+Recursive: 10946
+PROFILE: interrupts/evictions/bytes = 4/0/608
 $ ls
-main    main.cpp
-$ ./main
-Inside main
-Inside func1
-Inside new_func1
-Inside func2
-$ ls
-gmon.out    main    main.cpp
+fibonnaci    fibonnaci.cpp    gperftool.prof
 ~~~
 
-Now that we have `gmon.out`, we can ask `gprof` to show us the profile.
+Now that we have `gperftool.prof`, we will use a program called `pprof` to turn this data into something `kcachegrind` can understand.
+`pprof` needs to know both the executable you ran (`./fibonnaci`) and the file holding `gperftool`'s profile output (`gperftool.prof`).
+We'll store this in a file named `gperftool.out`:
 
 ~~~shell
-$ gprof main
+$ pprof --callgrind ./fibonnaci gperftool.prof > gperftool.out
 ~~~
 
+Last but not least, open `gperftool.out` in `kcachegrind`:[^X11-dont-forget]
+
+~~~shell
+$ kcachegrind gperftool.out
+~~~
+
+To review, to use `gperftool`, you need to do the following:
+
+1. Link the profiler library into your program with `-lprofiler` and enable debug information with `-g`
+2. Run your program with the `CPUPROFILE` environment variable set to the name of the profiler output file
+3. Use `pprof` to convert the profiler output file into a format `kcachegrind` understands
+4. Use `kcachegrind` to visualize your output!
+
+### Using `kcachegrind` to Interpret `gperftool` Statistics
+
+When you open `kcachegrind`, you will see three panes.
+On the left is the **flat profile**, which shows a list of functions and statistics about them.
+On the bottom right is the **call graph** --- a directed graph showing what functions each function calls.
```

Comment:
By default this window opens with the callees tab but the call graph tab is right next to it so this might not be an issue, but none the less could be slightly confusing.

Reply:
Good note. `kcachegrind` remembers what you had open last time, so I have no idea how it looks on a fresh install.